### PR TITLE
Improve async workers to handle forking for async compile

### PR DESCRIPTION
**Pull request by aorenste**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141659
* __->__ #142072

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/async_compile.py [hunk 1]
```diff
@@ -357,3 +358,10 @@ def wait(self, scope: Dict[str, Any]) -> None:
     pass
 else:
     AsyncCompile.warm_pool()
+
```
Comment (masnesral) [inline torch/_inductor/async_compile.py:361]: So was this just missing before?

File: torch/_inductor/compile_worker/subproc_pool.py [hunk 2]
```diff
@@ -157,31 +187,34 @@ def submit(
...
     def _read_thread(self) -> None:
-        try:
-            while True:
-                job_id, data = _recv_msg(self.read_pipe)
-                if job_id < 0:
-                    if self.running:
-                        log.warning("SubprocPool unclean exit")
-                    self.read_pipe.close()
+        while True:
+            job_id, data = _recv_msg(self.read_pipe)
```
Comment (masnesral) [inline torch/_inductor/compile_worker/subproc_pool.py:191]: Are you sure we no longer need a try around the recv? I'd assume that could raise if the pipe on the other end closes, for example

File: torch/_inductor/compile_worker/subproc_pool.py [hunk 3]
```diff
@@ -273,14 +317,18 @@ def callback(_: Future[Any]) -> None:
         future.add_done_callback(callback)
 
     @staticmethod
-    def do_job(data: bytes) -> bytes:
+    def do_job(pickler: SubprocPickler, data: bytes) -> bytes:
         # do the pickle/unpickle in the sub-subproc
-        job = pickle.loads(data)
+        shape_env = ShapeEnv()
```
Comment (masnesral) [inline torch/_inductor/compile_worker/subproc_pool.py:322]: I have no idea what's happening here :/

File: torch/fx/_graph_pickler.py [hunk 4]
```diff
@@ -565,3 +565,13 @@ def unpickle(self, unpickle_state: _UnpickleState) -> TracingContext:
             self.force_unspec_int_unbacked_size_like
         )
         return context
+
+
+class SubprocGraphPicker(torch._inductor.compile_worker.subproc_pool.SubprocPickler):
+    @override
```
Comment (masnesral) [inline torch/fx/_graph_pickler.py:571]: comment about what this is?

**Review by masnesral — CHANGES_REQUESTED** [review]
LG; I'm not sure about the FakeTensor part. If that's intentional, perhaps a comment?

File: torch/_inductor/async_compile.py (hunk 1 above)
Comment (aorenste) [reply torch/_inductor/async_compile.py:361]: I think so. It only happened sometimes (when the child exited uncleanly?) so not normally seen.

File: torch/_inductor/compile_worker/subproc_pool.py (hunk 2 above)
Comment (aorenste) [reply torch/_inductor/compile_worker/subproc_pool.py:191]: I looked at it closer. I think we're both right - with the except in the place it was before then there was a problem with exceptions being caught and ignored. In the next version I've added a try around the recv which then notifies the future about the exception.

File: torch/_inductor/compile_worker/subproc_pool.py (hunk 3 above)
Comment (aorenste) [reply torch/_inductor/compile_worker/subproc_pool.py:322]: Oh - that leaked in from a future PR - I'll move it where it's needed.

File: torch/fx/_graph_pickler.py (hunk 4 above)
Comment (aorenste) [reply torch/fx/_graph_pickler.py:571]: Another bit of code that belongs in a future PR.  Moved.

**Review by masnesral — APPROVED** [review]

**Pull request closed without merging**